### PR TITLE
rental and intercity flow

### PR DESCRIPTION
**Pull request by max-keviv**
## Type of Change

- [ ] Bugfix
- [x] New feature
- [ ] Enhancement
- [ ] Refactoring
- [ ] Dependency updates

## Description
Driver Rental Flow UI changes

1. added support for end ride otp for rental ride
2. odometer reading and image input for rental ride
3. show different new ride popup for rental
4. add rental ride polling every minute if there is destination location change then update the destination
5. added the add new stop callback on new stop FCM

figma - https://www.figma.com/file/ZDDROi2X7jZlhiRF6k9Ubl/YS-Intercity%2B-Rental---Presentation?node-id=0%3A1&mode=dev

API Changes needed

start ride api - new field for start odometer reading and image

end ride api - new field for end odometer reading and image

rideHistory - it will have field optional type rideScheduleTime, startOdometerReading, endOdometerReading and startOdometerImage and endOdometerImage and rideProductType (which will help to identify the ride type) and rideStartTime

add stop apis

New Add Stop FCM

### Additional Changes

- [ ] This PR modifies the database schema (database migration added)
- [ ] This PR modifies dhall configs/environment variables
- [ ] This PR contains API breaking changes

## Checklist

- [ ] I formatted the code and addressed linter errors `./dev/format-all-files.sh`
- [ ] I reviewed submitted code
- [ ] I added unit tests for my changes where possible
- [ ] I added a [CHANGELOG](/CHANGELOG.md) entry if applicable
- [ ] No leak detected in leakcanary

### CODE REVIEW

File: Frontend/ui-driver/src/Screens/RideSelectionScreen/Controller.purs [hunk 1]
```diff
@@ -167,7 +167,7 @@ rideHistoryListTransformer list categoryAction =
     , source : toPropValue (decodeAddress (ride.fromLocation) false)
     , updatedAt  : toPropValue ride.updatedAt
     , driverName : toPropValue ride.driverName
-    , destination  : toPropValue (decodeAddress (ride.toLocation) false)
+    , destination  : toPropValue (maybe "" (\toLocation -> decodeAddress (toLocation) false) ride.toLocation)
```
Comment (witcher-shailesh) [inline Frontend/ui-driver/src/Screens/RideSelectionScreen/Controller.purs:170]: what's the reason for this change?

File: Frontend/ui-driver/src/Screens/ReferralScreen/View.purs [hunk 2]
```diff
@@ -295,7 +295,7 @@ noDataView state =
             [ width MATCH_PARENT
             , height WRAP_CONTENT
             , margin (Margin 24 0 24 30)
-            , text (getString PLEASE_WAIT_WHILE_WE_UPDATE_THE_DETAILS)
+            , text (getStringEnToHi PLEASE_WAIT_WHILE_WE_UPDATE_THE_DETAILS)
```
Comment (witcher-shailesh) [inline Frontend/ui-driver/src/Screens/ReferralScreen/View.purs:298]: why are this strings added as EnToHi?

Comment (witcher-shailesh) [reply Frontend/ui-driver/src/Screens/ReferralScreen/View.purs:298]: we have this workflow working for all languages

**Review by witcher-shailesh — CHANGES_REQUESTED** [review]

File: Frontend/ui-driver/src/Screens/HomeScreen/View.purs [hunk 3]
```diff
@@ -1860,6 +1903,49 @@ rideStatusPolling pollingId duration state push action = do
       Left err -> pure unit
     else pure unit
 
+rentalRideStatusPolling :: forall action. String -> Number -> HomeScreenState -> (action -> Effect Unit) -> (Paths -> Maybe StopLocation -> Maybe StopLocation -> action) -> Flow GlobalState Unit
+rentalRideStatusPolling pollingId duration state push action = do
+ 
+  if isRentalRideStatusPollingActive state then do
+    _ <- pure $ spy "inside rentalRideStatusPolling" ""
```
Comment (witcher-shailesh) [inline Frontend/ui-driver/src/Screens/HomeScreen/View.purs:1910]: remove spy/log statements

File: Frontend/ui-driver/src/Screens/HomeScreen/Controller.purs [hunk 4]
```diff
@@ -479,8 +514,11 @@ eval BackPressed state = do
     continue state
   else if state.props.enterOtpModal then continue state { props = state.props { rideOtp = "", enterOtpFocusIndex = 0, enterOtpModal = false, rideActionModal = true } }
   else if (state.props.currentStage == ST.ChatWithCustomer) then do
-    _ <- pure $ setValueToLocalStore LOCAL_STAGE (show ST.RideAccepted)
-    continue state{props{currentStage = ST.RideAccepted}}
+    previousStage <- pure $ getValueToLocalStore PREVIOUS_LOCAL_STAGE
+    _ <- pure $ setValueToLocalStore PREVIOUS_LOCAL_STAGE $ show state.props.currentStage
```
Comment (witcher-shailesh) [inline Frontend/ui-driver/src/Screens/HomeScreen/Controller.purs:518]: instead of using `_ <-` try to use `void $`

File: Frontend/ui-driver/src/Screens/HomeScreen/Controller.purs [hunk 5]
```diff
@@ -507,7 +545,9 @@ eval BackPressed state = do
     continue state
 
 eval TriggerMaps state = continueWithCmd state[ do
-  _ <- pure $ openNavigation state.data.activeRide.dest_lat state.data.activeRide.dest_lon "DRIVE"
+  let destLat = if (state.data.activeRide.tripType == ST.Rental) then fromMaybe state.data.activeRide.src_lat state.data.activeRide.nextStopLat else state.data.activeRide.dest_lat
```
Comment (witcher-shailesh) [inline Frontend/ui-driver/src/Screens/HomeScreen/Controller.purs:548]: in case it's rental ride and nextStopLat is Nothing, we should not use src_lat of the active ride and start the navigation. This should be handled properly

File: Frontend/ui-driver/src/Screens/HomeScreen/Controller.purs
Comment (witcher-shailesh) [inline Frontend/ui-driver/src/Screens/HomeScreen/Controller.purs:1]: remove unnecessary logs

File: Frontend/ui-driver/src/Resource/Constants.purs [hunk 6]
```diff
@@ -96,3 +98,61 @@ waitTimeConstructor key = case key of
...
+getHomeStageFromString :: String -> ST.HomeScreenStage
+getHomeStageFromString localStage = 
+  case localStage of
+        "HomeScreen" -> ST.HomeScreen
+        "RideRequested" -> ST.RideRequested
+        "RideAccepted" -> ST.RideAccepted
+        "RideStarted" -> ST.RideStarted
+        "RideCompleted" -> ST.RideCompleted
+        "ChatWithCustomer" -> ST.ChatWithCustomer
+        _ -> ST.HomeScreen
```
Comment (witcher-shailesh) [inline Frontend/ui-driver/src/Resource/Constants.purs:158]: we have multiple screen stages for homeScreen, why rest of them are default to HomeScreen, this can cause unexpected behaviors in future

Comment (witcher-shailesh) [reply Frontend/ui-driver/src/Resource/Constants.purs:158]: also where are we using it as String? why not directly use the constructors?

File: Frontend/ui-common/src/Components/RideCompletedCard/View.purs [hunk 7]
```diff
@@ -755,4 +760,157 @@ lottieQRView config push =
...
+
+infoCardView :: forall w. Config -> String -> InfoCardConfig -> PrestoDOM (Effect Unit) w
+infoCardView config orientation' infoCardConfig = 
+  linearLayout
+  [ height infoCardConfig.height
+  , weight 1.0
+  , width MATCH_PARENT
+  , margin infoCardConfig.margin
+  , gravity LEFT
+  , orientation if orientation' == "VERTICAL" then VERTICAL else HORIZONTAL
```
Comment (witcher-shailesh) [inline Frontend/ui-common/src/Components/RideCompletedCard/View.purs:871]: use boolToVisibility from Mobility.Prelude module

File: Frontend/android-native/app/ysDriverProd/debug/output-metadata.json
Comment (witcher-shailesh) [inline Frontend/android-native/app/ysDriverProd/debug/output-metadata.json:1]: remove this file from the PR

**Review by witcher-shailesh — CHANGES_REQUESTED** [review]

File: Frontend/ui-driver/src/Services/Backend.purs [hunk 8]
```diff
@@ -294,17 +298,21 @@ makeStartRideReq otp lat lon ts = StartRideReq {
...
+      unwrapResponse (x) = x
+
 
-makeEndRideReq :: Number -> Number -> Maybe Boolean -> Int -> Int -> String -> EndRideReq
-makeEndRideReq lat lon numDeviation tripDistance tripDistanceWithAcc ts = EndRideReq {
+makeEndRideReq :: Maybe String -> Maybe String -> Maybe String -> Number -> Number -> Maybe Boolean -> Int -> Int -> String -> EndRideReq
+makeEndRideReq endOtp endOdometerReading fileId lat lon numDeviation tripDistance tripDistanceWithAcc ts = EndRideReq {
+    "endRideOtp" : endOtp,
+    "odometer" : (\value -> Odometer {
+        "value" : fromMaybe 0.0 $ Number.fromString value,
```
[the remark this answers is not in the document]
Comment (max-keviv) [reply Frontend/ui-driver/src/Services/Backend.purs:313]: we are doing that only, if endOdometerReading is nothing we are sending nothing,
here we are using converting it `fromMaybe 0.0 $ Number.fromString value,` 
should we make "odometer" nothing if fromString gives nothing ?

[the remark this answers is not in the document]
Comment (witcher-shailesh) [reply Frontend/ui-driver/src/Services/Backend.purs:313]: I think it's like this because we are calling `fromString` which again returns maybe, considering this we can re-write this where 0.0 should not go even if `fromString` function fails. But, It's okay for now, we can look into major fixes majorly

File: Frontend/ui-driver/src/Services/API.purs [hunk 9]
```diff
@@ -542,9 +585,80 @@ newtype RidesInfo = RidesInfo
...
+  }
+
+derive instance genericOdometerReading :: Generic OdometerReading _
+derive instance newtypeOdometerReading :: Newtype OdometerReading _
+instance standardEncodeOdometerReading :: StandardEncode OdometerReading where standardEncode (OdometerReading req) = standardEncode req
+instance showOdometerReading :: Show OdometerReading where show = genericShow
+instance decodeOdometerReading :: Decode OdometerReading where decode = defaultDecode
+instance encodeOdometerReading :: Encode OdometerReading where encode = defaultEncode
+
+newtype StopLocationAddress = StopLocationAddress
```
[the remark this answers is not in the document]
Comment (max-keviv) [reply Frontend/ui-driver/src/Services/API.purs:614]: in driver side we have type LocationInfo in which lat and lon is mandatory here in type StopLocationAddress we don't have lat lon

File: Frontend/ui-driver/src/Screens/RideSelectionScreen/Controller.purs (hunk 1 above)
Comment (max-keviv) [reply Frontend/ui-driver/src/Screens/RideSelectionScreen/Controller.purs:170]: toLocation is now optional field, we don't show tolocation in the ridehistory and invoice for rentals

File: Frontend/ui-driver/src/Screens/ReferralScreen/View.purs (hunk 2 above)
Comment (max-keviv) [reply Frontend/ui-driver/src/Screens/ReferralScreen/View.purs:298]: this was the requirement to show in hindi only if merchant is yatrisathi cc: @aryan-13

File: Frontend/ui-common/src/Components/RideCompletedCard/View.purs (hunk 7 above)
Comment (max-keviv) [reply Frontend/ui-common/src/Components/RideCompletedCard/View.purs:871]: we are not setting visibility

File: Frontend/ui-driver/src/Resource/Constants.purs (hunk 6 above)
Comment (max-keviv) [reply Frontend/ui-driver/src/Resource/Constants.purs:218]: we are storing the previous homeScreenStage in localstorage and here I was making constructor from string
should i match directly string ?

File: Frontend/android-native/mobility-app/src/main/java/in/juspay/mobility/app/OverlayMessagingService.java [hunk 10]
```diff
@@ -273,12 +278,27 @@ private void setViewListeners(View messageView) {
...
 
+    public void openNavigation(double lat, double lon) {
+        try {
+            Uri mapsURI = Uri.parse("google.navigation:q=" + lat + "," + lon);
+            Intent mapIntent = new Intent(Intent.ACTION_VIEW, mapsURI);
+            mapIntent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
+            mapIntent.setPackage("com.google.android.apps.maps");
+            startActivity(mapIntent);
+        } catch (Exception e) {
+            e.printStackTrace();
```
Comment (Rohit4997) [inline Frontend/android-native/mobility-app/src/main/java/in/juspay/mobility/app/OverlayMessagingService.java:289]: Add firebase log here to track this exception

File: Frontend/android-native/mobility-app/src/main/java/in/juspay/mobility/app/OverlaySheetService.java [hunk 11]
```diff
@@ -142,7 +148,74 @@ private void updateTagsView (SheetAdapter.SheetViewHolder holder, SheetModel mod
             String pickupChargesText = formattedPickupChargesText;
             String searchRequestId = model.getSearchRequestId();
             boolean showVariant =  !model.getRequestedVehicleVariant().equals(NO_VARIANT) && model.isDowngradeEnabled() && RideRequestUtils.handleVariant(holder, model, this);
-            if (model.getCustomerTip() > 0 || model.getDisabilityTag() || searchRequestId.equals(DUMMY_FROM_LOCATION) || model.isGotoTag() || showVariant || showSpecialLocationTag) {
+            if(model.getRideProductType().equals(RENTAL)){
```
Comment (Rohit4997) [inline Frontend/android-native/mobility-app/src/main/java/in/juspay/mobility/app/OverlaySheetService.java:151]: Create A different function `updateRentalData` in RideReqUtils and use it across both OverlaySheetService and RideReqActivity

You can Refer to `updateTierAndAC`

File: Frontend/android-native/mobility-app/src/main/java/in/juspay/mobility/app/OverlaySheetService.java [hunk 12]
```diff
@@ -142,7 +148,74 @@ private void updateTagsView (SheetAdapter.SheetViewHolder holder, SheetModel mod
...
+                holder.tagsBlock.setVisibility(View.VISIBLE);
+                holder.vcTierAndACView.setVisibility(View.GONE);
+                holder.reqButton.setTextColor(getColor(R.color.white));
+                holder.reqButton.setBackgroundTintList(ColorStateList.valueOf(getColor(R.color.turquoise)));
+                if (!variant.equals(NO_VARIANT) && key.equals("yatrisathiprovider")) {
```
Comment (Rohit4997) [inline Frontend/android-native/mobility-app/src/main/java/in/juspay/mobility/app/OverlaySheetService.java:156]: Why this variant change is required 
I Can't see it in design

File: Frontend/android-native/mobility-app/src/main/java/in/juspay/mobility/app/OverlaySheetService.java [hunk 13]
```diff
@@ -142,7 +148,74 @@ private void updateTagsView (SheetAdapter.SheetViewHolder holder, SheetModel mod
...
+                       holder.rideTypeTag.setBackgroundResource(R.drawable.ic_orange_tag);
+                        holder.rideTypeImage.setVisibility(View.GONE);
+                    }
+                    holder.rideTypeText.setText(variant);
+                }
+                holder.rentalRideTypeTag.setVisibility(View.VISIBLE);
+                holder.rideStartDateTimeTag.setVisibility(View.VISIBLE);
+                holder.rideStartTime.setText(model.getRideStartTime());
+                holder.rideStartDate.setVisibility(View.VISIBLE);
+                if(model.getRideStartDate() != "")
```
Comment (Rohit4997) [inline Frontend/android-native/mobility-app/src/main/java/in/juspay/mobility/app/OverlaySheetService.java:171]: `equals()` Use this function 
check this change everywhere in this file

File: Frontend/android-native/mobility-app/src/main/java/in/juspay/mobility/app/MyFirebaseMessagingService.java [hunk 14]
```diff
@@ -428,6 +443,37 @@ public void startFCMBundleUpdateService(RemoteMessage remoteMessage, String merc
         }
     }
 
+ private void addUpdateStop(JSONObject driverPayloadJsonObject){
+        try {
+            String stopTitle = driverPayloadJsonObject.has("isEdit") && driverPayloadJsonObject.getBoolean("isEdit") ? getString(R.string.customer_edited) : getString(R.string.customer_added); // TODO:: Temporary added until this data comes from backend
```
[the remark this answers is not in the document]
Comment (max-keviv) [reply Frontend/android-native/mobility-app/src/main/java/in/juspay/mobility/app/MyFirebaseMessagingService.java:448]: nope

File: Frontend/android-native/mobility-app/src/main/java/in/juspay/mobility/app/OverlaySheetService.java (hunk 12 above)
Comment (max-keviv) [reply Frontend/android-native/mobility-app/src/main/java/in/juspay/mobility/app/OverlaySheetService.java:156]: this is for yatrisathi specific to show ac and non ac

**Review by witcher-shailesh — APPROVED** [review]
